tool/pftool.cc

```diff
@@ -780,6 +879,207 @@ static void daq( const std::string& cmd, PolarfireTarget* pft ) {
+    
+    printf(" Clearing charge injection on all channels (ground-state)...\n");
+    
+    /// first, disable charge injection on all channels
+    for (int ilink=0; ilink<pft->hcal.elinks().nlinks(); ilink++) {
+      if (!pft->hcal.elinks().isActive(ilink)) continue;
+
+      int iroc=ilink/2;        
+      for (int ichan=0; ichan<NUM_ELINK_CHAN; ichan++) {
+        char pagename[32];
+        snprintf(pagename,32,"CHANNEL_%d",(ilink%2)*(NUM_ELINK_CHAN)+ichan);
+        // set the value
+        pft->hcal.roc(iroc).applyParameter(pagename, "LOWRANGE", 0);
+        pft->hcal.roc(iroc).applyParameter(pagename, "HIGHRANGE", 0);
+      }
+    }
+
+    printf(" Enabling IntCTest...\n");
+    
+    for (int ilink=0; ilink<pft->hcal.elinks().nlinks(); ilink++) {
+      if (!pft->hcal.elinks().isActive(ilink)) continue;
+
+      int iroc=ilink/2;        
+      char pagename[32];
+      snprintf(pagename,32,"REFERENCE_VOLTAGE_%d",(ilink%2));
+      // set the value
+      pft->hcal.roc(iroc).applyParameter(pagename, "INTCTEST", 1);
+      if (cmd=="SCANCHARGE") {
+        snprintf(pagename,32,"DIGITAL_HALF_%d",(ilink%2));
+        // set the value
+        pft->hcal.roc(iroc).applyParameter(pagename, "L1OFFSET", 8);
+      }
+    }
+    
+    ////////////////////////////////////////////////////////////
+    
+    for (int step=0; step<steps; step++) {
+
+      ////////////////////////////////////////////////////////////
+      /// set values
+      int value=low_value+step*(high_value-low_value)/std::max(1,steps-1);
+
+      printf(" Scan %s=%d...\n",valuename.c_str(),value);
+      
+      for (int ilink=0; ilink<pft->hcal.elinks().nlinks(); ilink++) {
+        if (!pft->hcal.elinks().isActive(ilink)) continue;
+
+        int iroc=ilink/2;
+        char pagename[32];
+        snprintf(pagename,32,pagetemplate.c_str(),ilink%2);
+        // set the value
+        pft->hcal.roc(iroc).applyParameter(pagename, valuename, value);
+      }
+      ////////////////////////////////////////////////////////////
+
+      pft->prepareNewRun();
+      
+      for (int ichan=0; ichan<NUM_ELINK_CHAN; ichan++) {
+
+        ////////////////////////////////////////////////////////////
+        /// Enable charge injection channel by channel -- per elink
+        for (int ilink=0; ilink<pft->hcal.elinks().nlinks(); ilink++) {
+          if (!pft->hcal.elinks().isActive(ilink)) continue;
+          
+          int iroc=ilink/2;        
+          char pagename[32];
+          snprintf(pagename,32,"CHANNEL_%d",(ilink%2)*(NUM_ELINK_CHAN)+ichan);
+          // set the value
+          pft->hcal.roc(iroc).applyParameter(pagename, modeinfo, 1);
+        }
+                
+        //////////////////////////////////////////////////////////
+        /// Take the expected number of events and save the events
+        for (int ievt=0; ievt<events_per_step; ievt++) {
+
+          pft->backend->fc_calibpulse();
+          std::vector<uint32_t> event = pft->daqReadEvent();
```

Comment:
This effectively limits the scan to only being done via non-DMA readout. Does this function even if DMA readout is enabled? If not, we should add a disclaimer to the description in the menu.